ayeletK/assignment2/Q2_solution.cpp

```diff
@@ -0,0 +1,71 @@
+#include "tree_node.cpp"
+#include <string>
+
+using namespace std;
+
+#define KEY_NOT_IN_PATH "-"
+#define EMPTY_PATH ""
+#define LEFT_SON '0'
+#define RIGHT_SON '1'
+
+template <typename Key>
+char get_addition_to_path(Node<Key>* son_node, Node<Key>* parent_node){
+  return parent_node->left == son_node ? LEFT_SON : RIGHT_SON;
+}
+
+template <typename Key>
+string get_path_string(Key key, Node<Key>* curr_node, Node<Key>* parent_node = NULL){
+  if (!curr_node) return KEY_NOT_IN_PATH;
+  Key curr_key = curr_node->get_key();
+  // EMPTY_PATH represents the case the root node is the key
+  if (curr_key == key){
+    if (!parent_node) return EMPTY_PATH;
+    else return string(1, get_addition_to_path(curr_node, parent_node));
+  }
+  string res;
+  res = get_path_string(key, curr_node->left, curr_node);
+  if (res == KEY_NOT_IN_PATH){
+    res = get_path_string(key, curr_node->right, curr_node);
+  }
+  if (res != KEY_NOT_IN_PATH && parent_node){
+    char addition = get_addition_to_path(curr_node, parent_node);
+    res.insert(res.begin(), addition);
+  }
+  return res;
+}
+
+template <typename Key>
+Node<Key>* get_node_from_path(Node<Key>* root, string path){
+  if (!root) throw "the given tree is empty.";
+  if (path == KEY_NOT_IN_PATH) return NULL;
+  if (path == EMPTY_PATH) return root;
+  Node<Key>* curr_node = root;
+  for (char c: path){
+    if (c == LEFT_SON) curr_node = curr_node->left;
+    else curr_node = curr_node->right;
+    if (!curr_node) throw "Illegal path";
+  }
+  return curr_node;
+}
+
+template <typename Key>
+Node<Key>* get_lowest_common_ancestor(Node<Key>* root, Node<Key>* node1, Node<Key>* node2){
```

Comment:
Nice solution: Find the paths from the root to each node; then return the node in which the paths diverge. :+1: 

A similar solution could use the nodes path from the root to each target node (instead of left/right, the list of node keys on the path, starting from the root). This could even use the function from previous solution which finds all ancestors :)

Reply:
I wasn't sure about my solution.
We weren't allowed to save the nodes in a different data structure. I wasn't sure if saving the keys is legal or not. Furthermore, The tree in Q2 doesn't have to be binary tree - so I didn't know if using the code from Q1 solution will useful in this case.
That's why I chose to implement with string representing the path. In that case, we can expend the solution of Q2 to non binary tree, without any need to change the solution for Q1.